src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -327,6 +384,157 @@ public void ConnectHeaders_SupplyHeaders_ToDownloadArraySizeSameAsNumberOfHeader
             Assert.Equal(headersToDownloadCount, peer2Headers.Count);
         }
 
+        /// <summary>
+        /// Issue 8 @ Create chained header tree component #1321
+        /// We have a chain and peer B is on the best chain.
+        /// After that peer C presents an alternative chain with a forkpoint below B's Tip.
+        /// After that peer A prolongs C's chain (but sends all headers from the fork point) with a few valid and one invalid.
+        /// Make sure that C's chain is not removed - only headers after that.
+        /// </summary>
+        [Fact]
+        public void ConnectHeaders_MultiplePeersWithForks_CorrectTip()
+        {
+            TestContext testContext = new TestContextBuilder().Build();
+            ChainedHeaderTree chainedHeaderTree = testContext.ChainedHeaderTree;
+            ChainedHeader chainTip = testContext.ExtendAChain(5);
+            chainedHeaderTree.Initialize(chainTip, true);
+            List<BlockHeader> listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 5);
+
+            chainedHeaderTree.ConnectNewHeaders(2, listOfExistingHeaders);
+            ChainedHeader peerTwoTip = chainedHeaderTree.GetChainedHeaderByPeerId(2);
+
+            // Peer 2 is on the best chain.
+            // b1 = b2 = b3 = b4 = b5
+            peerTwoTip.HashBlock.Should().BeEquivalentTo(chainTip.HashBlock);
+            peerTwoTip.ChainWork.Should().BeEquivalentTo(chainTip.ChainWork);
+
+            // Peer 3 presents an alternative chain with a forkpoint below Peer 2's Tip.
+            // b1 = b2 = b3 = b4 = b5
+            //         = c3 = c4 = c5
+            ChainedHeader peerThreeTip = testContext.ExtendAChain(3, chainTip.Previous.Previous.Previous);
+            List<BlockHeader> listOfNewHeadersFromPeerThree = testContext.ChainedHeaderToList(peerThreeTip, 3);
+            chainedHeaderTree.ConnectNewHeaders(3, listOfNewHeadersFromPeerThree);
+            
+            ChainedHeader fork = chainTip.FindFork(peerThreeTip);
+            fork.Height.Should().BeLessThan(peerTwoTip.Height);
+
+            // Peer 1 prolongs Peer 3's chain with one invalid block.
+            // c3 = c4 = c5 = |a6| = a7 = a8
+            const int numberOfBlocksToExtend = 3;
+            ChainedHeader peerOneTip = testContext.ExtendAChain(numberOfBlocksToExtend, peerThreeTip);
+            List <BlockHeader> listOfPeerOnesHeaders = testContext.ChainedHeaderToList(peerOneTip, numberOfBlocksToExtend);
+            
+            int depthOfInvalidHeader = 2;
+            BlockHeader invalidBlockHeader = listOfPeerOnesHeaders[depthOfInvalidHeader];
```

Comment:
Can we make it in a way that the invalid block is a7? This would be nice because we would test that blocks before as well as after the invalid block were not added to the tree.

Reply:
Done